headjack/models/utterance.py

```diff
@@ -49,6 +42,27 @@ def convo(
         history = history[::-1]
         return "\n".join(str(u) for u in history) + "\n"
 
+    @property
+    def log_str(self) -> str:
```

Comment:
Can this just be the `__str__` method and then the `log` method below won't be required? Then it'll be just like a simple logging event.
```py
all_utterances = [response] + response.history()
for utterance in all_utterances if not utterance._logged:
    _logger.info(utterance)
```

Reply:
@samredai so anywhere we want to log we'll have to put such a loop?

Reply:
@samredai so anywhere we want to log we'll have to put such a loop?

Reply:
It's more of a style nit but I'd rather have 3 explicit lines repeated in a number of places than an opaque call to a class method. For only 2 extra lines it will be obvious at first glance that the code is a call to the standard python logger.